File: server/src/main/java/cn/iinti/atom/system/LoginInterceptor.java
package cn.iinti.atom.system;


import cn.iinti.atom.BuildConfig;
import cn.iinti.atom.entity.CommonRes;
import cn.iinti.atom.entity.UserInfo;
import cn.iinti.atom.service.base.UserInfoService;
import com.alibaba.fastjson.JSONObject;
import com.google.common.base.Charsets;
import com.google.common.collect.Lists;
import jakarta.annotation.Resource;
import jakarta.servlet.http.Cookie;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import org.apache.commons.lang3.BooleanUtils;
import org.apache.commons.lang3.StringUtils;
import org.springframework.lang.Nullable;
import org.springframework.stereotype.Component;
import org.springframework.web.method.HandlerMethod;
import org.springframework.web.servlet.HandlerInterceptor;
import org.springframework.web.servlet.ModelAndView;

import java.io.IOException;
import java.lang.reflect.Method;
import java.util.List;


@Component
public class LoginInterceptor implements HandlerInterceptor {
    @Resource
    private UserInfoService userInfoService;

    private static final byte[] needLoginResponse = JSONObject.toJSONString(CommonRes.failed("请登录后访问")).getBytes(Charsets.UTF_8);
    private static final byte[] loginExpire = JSONObject.toJSONString(CommonRes.failed("请重新登录")).getBytes(Charsets.UTF_8);
    private static final byte[] onlyForAdminResponse = JSONObject.toJSONString(CommonRes.failed("非管理员")).getBytes(Charsets.UTF_8);

    private List<String> collectTokenList(HttpServletRequest request) {
        List<String> tokenList = Lists.newArrayList();

        // header 不区分大小写
        String operatorToken = request.getHeader(BuildConfig.userLoginTokenKey);
        if (StringUtils.isNotBlank(operatorToken)) {
            tokenList.add(operatorToken);
        }
        operatorToken = request.getParameter(BuildConfig.userLoginTokenKey);
        if (StringUtils.isNotBlank(operatorToken)) {
            tokenList.add(operatorToken);
        }

        Cookie[] cookies = request.getCookies();
        if (cookies != null) {
            for (Cookie cookie : cookies) {
                if (cookie.getName().equals(BuildConfig.userLoginTokenKey)) {
                    tokenList.add(cookie.getValue());
                }
            }
        }
        return tokenList.stream()
                .filter(s -> StringUtils.isNotBlank(s) &&
                        // undefined,null as blank from frontend javascript
                        !StringUtils.containsAny(s, "undefined", "null"))
                .toList();
    }

    private boolean handleNoToken(LoginRequired loginRequired, HttpServletResponse response) throws IOException {
        // 不存在鉴权token
        if (loginRequired == null) {
            //不需要登陆
            return true;
        }
        // 需要登录，但是没有token
        response.addHeader("content-type", "application/json; charset=utf-8");
        response.getOutputStream().write(needLoginResponse);
        return false;
    }

    @Override
    public boolean preHandle(@Nullable HttpServletRequest request, @Nullable HttpServletResponse response, @Nullable Object handler) throws Exception {
        if (!(handler instanceof HandlerMethod)) {
            return true;
        }

        if (request == null || response == null) {
            return true;
        }

        Method method = ((HandlerMethod) handler).getMethod();
        LoginRequired loginRequired = method.getAnnotation(LoginRequired.class);
        List<String> tokenList = collectTokenList(request);
        if (tokenList.isEmpty()) {
            return handleNoToken(loginRequired, response);
        }

        CommonRes<UserInfo> result = userInfoService.checkLogin(tokenList);
        if (loginRequired == null) {
            // no need login
            if (result.isOk()) {
                // but user send userToken
                AppContext.setUser(result.getData());
            }
            return true;
        }

        if (!result.isOk()) {
            //如果这个接口允许api token访问，那么直接允许，并且使用对应的token所在账户身份
            if (loginRequired.apiToken()) {
                result = userInfoService.checkAPIToken(tokenList);
            }
            if (!result.isOk()) {
                response.addHeader("content-type", "application/json; charset=utf-8");
                response.getOutputStream().write(loginExpire);
                return false;
            }
            AppContext.markApiUser();
        }
        if (loginRequired.forAdmin() && BooleanUtils.isNotTrue(result.getData().getIsAdmin())) {
            response.addHeader("content-type", "application/json; charset=utf-8");
            response.getOutputStream().write(onlyForAdminResponse);
            return false;
        }
        AppContext.setUser(result.getData());
        AppContext.setLoginAnnotation(loginRequired);
        return true;
    }

    @Override
    public void postHandle(@Nullable HttpServletRequest request, @Nullable HttpServletResponse response, @Nullable Object handler, ModelAndView modelAndView) {
        AppContext.removeUser();
    }
}

File: server/src/main/kotlin/cn/iinti/atom/system/LoginInterceptor.kt
package cn.iinti.atom.system

import cn.iinti.atom.BuildInfo
import cn.iinti.atom.entity.CommonRes.Companion.failed
import cn.iinti.atom.service.base.UserInfoService
import com.alibaba.fastjson.JSONObject
import jakarta.annotation.Resource
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.apache.commons.lang3.BooleanUtils
import org.apache.commons.lang3.StringUtils
import org.springframework.stereotype.Component
import org.springframework.web.method.HandlerMethod
import org.springframework.web.servlet.HandlerInterceptor
import org.springframework.web.servlet.ModelAndView
import java.io.IOException
import java.util.*

@Component
class LoginInterceptor : HandlerInterceptor {
    @Resource
    private lateinit var userInfoService: UserInfoService

    private fun collectTokenList(request: HttpServletRequest): List<String> {
        val tokenList: MutableList<String> = mutableListOf()
        // header 不区分大小写
        request.apply {
            getHeader(BuildInfo.userLoginTokenKey)?.apply {
                tokenList.add(this)
            }
            getParameter(BuildInfo.userLoginTokenKey)?.apply {
                tokenList.add(this)
            }
            cookies?.find { it.name == BuildInfo.userLoginTokenKey }
                ?.apply { tokenList.add(value) }
        }

        return tokenList
            .filter { it.isNotBlank() }
            // undefined,null as blank from frontend javascript
            .filter { !StringUtils.containsAny(it, "undefined", "null") }
            .toList()
    }

    @Throws(IOException::class)
    private fun handleNoToken(loginRequired: LoginRequired?, response: HttpServletResponse): Boolean {
        // 不存在鉴权token
        if (loginRequired == null) {
            //不需要登陆
            return true
        }
        // 需要登录，但是没有token
        response.addHeader("content-type", "application/json; charset=utf-8")
        response.outputStream.write(needLoginResponse)
        return false
    }

    @Throws(Exception::class)
    override fun preHandle(request: HttpServletRequest, response: HttpServletResponse, handler: Any): Boolean {
        if (handler !is HandlerMethod) {
            return true
        }

        val method = handler.method
        val loginRequired = method.getAnnotation(LoginRequired::class.java)
        val tokenList = collectTokenList(request)
        if (tokenList.isEmpty()) {
            return handleNoToken(loginRequired, response)
        }

        var result = userInfoService.checkLogin(tokenList)
        if (loginRequired == null) {
            // no need login
            if (result.isOk()) {
                // but user send userToken
                AppContext.setUser(result.data)
            }
            return true
        }

        if (!result.isOk()) {
            //如果这个接口允许api token访问，那么直接允许，并且使用对应的token所在账户身份
            if (loginRequired.apiToken) {
                result = userInfoService.checkAPIToken(tokenList)
            }
            if (!result.isOk()) {
                response.addHeader("content-type", "application/json; charset=utf-8")
                response.outputStream.write(loginExpire)
                return false
            }
            AppContext.markApiUser()
        }
        if (loginRequired.forAdmin && BooleanUtils.isNotTrue(result.data!!.isAdmin)) {
            response.addHeader("content-type", "application/json; charset=utf-8")
            response.outputStream.write(onlyForAdminResponse)
            return false
        }
        AppContext.setUser(result.data)
        AppContext.setLoginAnnotation(loginRequired)
        return true
    }

    override fun postHandle(
        request: HttpServletRequest,
        response: HttpServletResponse,
        handler: Any,
        modelAndView: ModelAndView?
    ) {
        AppContext.clean()
    }

    companion object {
        private val needLoginResponse = JSONObject.toJSONString(failed<Any>("请登录后访问")).toByteArray()
        private val loginExpire = JSONObject.toJSONString(failed<Any>("请重新登录")).toByteArray()
        private val onlyForAdminResponse = JSONObject.toJSONString(failed<Any>("非管理员")).toByteArray()
    }
}